optuna/study.py

```diff
@@ -191,6 +191,15 @@ def __init__(
         self.sampler = sampler or samplers.TPESampler()
         self.pruner = pruner or pruners.MedianPruner()
 
+        if (
+                isinstance(self.sampler, samplers.TPESampler) and
+                isinstance(self.pruner, pruners.HyperbandPruner)):
+            msg = (
+                "The algorithm of TPESampler and HyperbandPruner might behave in a different way "
+                "from the paper because the sampler uses all the trials including ones of "
+                "brackets other than that of currently running trial")
+            warnings.warn(msg, UserWarning)
```

Comment:
Should we log this warnings as we're doing with the other ones? I also noted that it's not entirely clear whether the "paper" refers to TPE or Hyperband (or both). Do you think it'd be possible to clarify it?

Reply:
> Should we log this warnings as we're doing with the other ones?

I want to hear the dev's opinion because I don't have any taste here.

> I also noted that it's not entirely clear whether the "paper" refers to TPE or Hyperband (or both). Do you think it'd be possible to clarify it?

Good catch, thank you. By "paper," I meant to refer Hyperband mainly.

Reply:
>> Should we log this warnings as we're doing with the other ones?

> I want to hear the dev's opinion because I don't have any taste here.

In oputna code, `warnings.warn()` method is usually used with the `DeprecationWarning` category. If the category is specified, the warning message isn't output by default. So, I think, emitting the same message using Optuna's logger is useful in such case. 
However, since `UserWarning` category is specified here, if we use Optuna logger too, the duplicate messages will be shown to users (IMO, it's a bit noisy). I think we can simply replace `warnings.warn(msg, UserWarning)` with `_logger.warning(msg)` to align to other Optuna code.

Reply:
Thanks for the discussion! @sile are you referring to the fact that the [`DeprecationWarning` is ignored by default](https://docs.python.org/3.8/library/warnings.html#warning-categories)? That might be one way to deal with it but IMO, this Python warning will only be raised once so the clutter is not that much of an issue and it might be better to let the user control the handling of Python warning. 🤔

Reply:
> are you referring to the fact that the DeprecationWarning is ignored by default?

Please see the highlighted sentence in the following image:
![image](https://user-images.githubusercontent.com/181413/71861290-32005180-3139-11ea-95ca-8c3fd5cdab26.png)
From: https://docs.python.org/3.8/library/warnings.html#warning-categories

> this Python warning will only be raised once so the clutter is not that much of an issue and it might be better to let the user control the handling of Python warning. 

That makes sense. I think it's a good option to switch the warning mechanism in this repository from Optuna logger to Python warning. It seems okay to leave this line of code as it is.

Reply:
**FYI** I found a code that emits a warning message via both `warnings.warn(..., UserWarning)` and `_logger.warn(..)`.

https://github.com/optuna/optuna/blob/0fcf76290cc38bd11fc6a8ef149fb006caa7a651/optuna/study.py#L326-L334

Reply:
Maybe I wasn't clear but I meant to both `warn` and log as shown above. It's not a strong opinion, just consistent.

Reply:
Oh, I misunderstood your intention. It also seems fine to use both `warnings.warn` and `_logger.warning` to keep consistency.

Reply:
Sorry, it was obviously not clear...

Reply:
Thank you for all your opinions.
I added `_logger.warning(...)`.